### ISSUE TITLE
Questions about using entropy model to compress feature 

### ISSUE THREAD

**Issue by 1364406834** [InterDigitalInc/CompressAI#267]
## Documentation

I have some features that I hope can be compressed and decompressed through the entropy model of compress AI, and dynamic code rate control can be achieved. I hope to make my features low-entropy by adding rate-distortion loss to my main program. But unfortunately, no matter how I adjust the lambda value in rate-distortion loss, I cannot change the size of the compressed feature. This is the result even if I compress the feature directly without training. Do you have any ideas?

**Comment by YodaEmbedding** [comment]
1. Did you train the model using the R-D loss with a *single* fixed lambda? Did you observe the rate/distortion changing during training? Are all the layers receiving the distortion gradients? Are all the encoder layers receiving the rate gradients?
2. Did you try training with a range of lambdas? e.g., $\lambda = 0$ should give a model with rate=0 and 0 dB PSNR. If this isn't the case, then something is wrong with the definition of the model or the training loop.

**Comment by 1364406834** [comment]
Thank you for such a quick reply. After testing, I found that when I set lambda to 0, I can still recover my pre-compression data almost accurately, unlike what you said. When I set the lambda to be very large, even 1000, it does not change the size of the compressed feature.
During training, I observed that the rate-distortion loss decreases normally and stabilizes after reaching a certain value. But the declining part is only the bpp loss part, not the mse loss part of distortion. In fact, distrotion loss has barely changed at all.

**Comment by 1364406834** [comment]
The above problem has been solved. However, I have a new question. I sent a 1*18*35*31*31 float32 all-zero matrix (2.4M) into the entropy model for compression, but found that the compression result was still about 400K. I think this is not a satisfactory result. Do you have any ideas?

**Comment by 1364406834** [comment]
The above problem has been solved. However, I have a new question. I sent a [1,18,35,31,31] float32 all-zero matrix (2.4M) into the entropy model for compression, but found that the compression result was still about 400K. I think this is not a satisfactory result. Do you have any ideas?

**Comment by YodaEmbedding** [comment]
- What architecture are you using? bmshj2018-factorized? (visualized below)
  ![2023-12-28  05-13-22  365](https://github.com/InterDigitalInc/CompressAI/assets/721196/4e29e2f9-a3fb-499c-b116-685cbaed89d9)
- Did you input the matrix as $x$ or as $y$?
  - If $x$, then ensure that the training dataset contains a few "low-frequency near-zero mean" examples.
  - If $y$, then what may be happening is that each channel's encoding distribution is centered at some $\mu \neq 0$, e.g., $\mu = 31$, but you've inputted data distributed with most of its values near $0$.

**Comment by 1364406834** [comment]
My architecture is similar to the one you gave, since my input is already a feature, so I input it as y. I removed channel 1 and used channel 18 as batch size. How can I solve this problem you mentioned? I heard others say that compressai's probability table is fixed and it seems that I need to update it. Does this view have anything to do with my problem?

**Comment by YodaEmbedding** [comment]
You should call `.update()` before running compress/decompress, as is done in [`compressai.utils.eval_model`](https://github.com/InterDigitalInc/CompressAI/blob/a4ae2eeef7bdb1b84ba076ac0d650b523f3fa882/compressai/utils/eval_model/__main__.py#L178C6-L178C6). This ensures that the runtime encoder/decoder uses the distributions that the model was trained to expect.

**Comment by 1364406834** [comment]
Thanks for your help! I have seen the solution you gave in other issues. After trying it it didn't work.
Fortunately, I have solved this problem by changing the encoding method of ans to rangecoder and changing the relevant source code. A relatively satisfactory result was achieved. Thank you again for your help! I will close this issue.
